multinsclient/client.go

```diff
@@ -0,0 +1,177 @@
+	// given workflow ID. If unset, DefaultHasher is used.
+	Hasher func(workflowID string, max int) int
+
+	// If true, enabled namespaces will create clients that won't connect until
+	// first used. Otherwise, enabled namespaces will have their clients eagerly
+	// attempt to connect to the server when the client is created.
+	ConnectLazy bool
+}
+
+// DefaultHasher uses FNV-1a hashing.
+func DefaultHasher(workflowID string, max int) int {
+	h := fnv.New32a()
+	h.Write([]byte(workflowID))
+	return int(h.Sum32()) % max
+}
+
+// New creates a new client from the given options.
+func New(options Options) (*Client, error) {
+	c := &Client{
+		hasher:                  options.Hasher,
+		enabledNamespaceClients: map[string]client.Client{},
+	}
+
+	// Use default hasher if none given
+	if c.hasher == nil {
+		c.hasher = DefaultHasher
+	}
+
+	// Close any created clients of this didn't complete successfully
+	success := false
+	defer func() {
+		if !success {
+			for _, client := range c.enabledNamespaceClients {
+				client.Close()
+			}
+		}
+	}()
+
+	// Create all non-disabled clients
+	seenNamespaces := make(map[string]bool, len(options.Namespaces))
+	for _, namespace := range options.Namespaces {
+		// We don't support default namespace name fallback, we require it to be set
+		if namespace.ClientOptions.Namespace == "" {
+			return nil, fmt.Errorf("all namespaces must have a namespace in the client options")
+		}
+		if _, exists := seenNamespaces[namespace.ClientOptions.Namespace]; exists {
+			return nil, fmt.Errorf("duplicate namespace %q", namespace.ClientOptions.Namespace)
+		}
+		seenNamespaces[namespace.ClientOptions.Namespace] = true
+		c.allNamespaces = append(c.allNamespaces, namespace.ClientOptions.Namespace)
+		if !namespace.Disabled {
+			c.enabledNamespaces = append(c.enabledNamespaces, namespace.ClientOptions.Namespace)
+			var err error
+			// TODO(cretz): Could alter options to have client options unrelated to
+			// namespaces and then use the new client.NewClientFromExisting introduced
+			// in https://github.com/temporalio/sdk-go/pull/881 to share connections
+			// across namespaces, but having multiple clients is mostly harmless.
+			if options.ConnectLazy {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.NewLazyClient(namespace.ClientOptions)
+			} else {
+				c.enabledNamespaceClients[namespace.ClientOptions.Namespace], err = client.Dial(namespace.ClientOptions)
+			}
+			if err != nil {
+				return nil, fmt.Errorf("failed creating client for namespace %q: %w", namespace.ClientOptions.Namespace, err)
+			}
+		}
+	}
+
+	// Must be at least one enabled
+	if len(c.enabledNamespaces) == 0 {
+		return nil, fmt.Errorf("no enabled namespaces")
+	}
+
+	success = true
+	return c, nil
+}
+
+// NamespaceFor selects a namespace and client for the given workflow ID. The
+// resulting namespace is always the same for the same workflow ID.
```

Comment:
> The resulting namespace is always the same for the same workflow ID

This is not true if onlyEnabled = true and there is disabled namespace.

Reply:
It's actually always true for _this_ client which has an immutable set of namespaces. I explain below how, after Reset, this could be different.